C error when passing wrong struct to a struct method with a generic type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 11.6, 20G165
Processor: 16 cpus, 64bit, little endian, Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/wes/dev/test-v
vmodules: /Users/wes/.vmodules
vroot: /Users/wes/dev/v
vexe: /Users/wes/dev/v/v
vexe mtime: 2022-01-17 06:27:26
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 65eb383.db48594

Git version: git version 2.30.1 (Apple Git-130)
Git vroot status: weekly.2022.02-41-gdb48594b
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 93fa86be

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg main.v`
```v
fn main() {
	mut obj := Obj{}
	obj.set<Bar>(Foo{})
}

struct Foo {}

struct Bar {
	Foo
}

struct Obj {}

fn (mut o Obj) set<T>(val T) {}

```

**What did you expect to see?**

a compilation error like "cannot use `Foo` as `Bar` in argument 1 to `set`"


**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/main.3667146868542244640.tmp.c:10976:34: error: passing 'main__Foo' (aka 'struct main__Foo') to parameter of incompatible type 'main__Bar' (aka 'struct main__Bar')
        main__Obj_set_T_main__Bar(&obj, (main__Foo){EMPTY_STRUCT_INITIALIZATION});
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/main.3667146868542244640.tmp.c:1824:72: note: passing argument to parameter 'val' here
VV_LOCAL_SYMBOL void main__Obj_set_T_main__Bar(main__Obj* o, main__Bar val);
                                                                       ^
1 warning and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1243,10 +1243,15 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				continue
 			}
 			if exp_arg_typ.has_flag(.generic) {
-				continue
+				if utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
+					concrete_types)
+				{
+					exp_arg_typ = utyp
+				} else {
+					continue
+				}
 			}
-			c.check_expected_call_arg(got_arg_typ, c.unwrap_generic(exp_arg_typ), node.language,
-				arg) or {
+			c.check_expected_call_arg(got_arg_typ, exp_arg_typ, node.language, arg) or {
 				// str method, allow type with str method if fn arg is string
 				// Passing an int or a string array produces a c error here
 				// Deleting this condition results in propper V error messages

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1255,12 +1255,14 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 					continue
 				}
 
-				if got_utyp := c.table.resolve_generic_to_concrete(got_arg_typ, method.generic_names,
-					concrete_types)
-				{
-					got_arg_typ = got_utyp
-				} else {
-					continue
+				if got_arg_typ.has_flag(.generic) {
+					if got_utyp := c.table.resolve_generic_to_concrete(got_arg_typ, method.generic_names,
+						concrete_types)
+					{
+						got_arg_typ = got_utyp
+					} else {
+						continue
+					}
 				}
 			}
 			c.check_expected_call_arg(got_arg_typ, exp_arg_typ, node.language, arg) or {

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1243,10 +1243,14 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				continue
 			}
 			if exp_arg_typ.has_flag(.generic) {
-				if utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
-					concrete_types)
-				{
-					exp_arg_typ = utyp
+				if concrete_types.len > 0 {
+					if utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
+						concrete_types)
+					{
+						exp_arg_typ = utyp
+					} else {
+						continue
+					}
 				} else {
 					continue
 				}

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1212,7 +1212,7 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 			}
 			exp_arg_sym := c.table.sym(exp_arg_typ)
 			c.expected_type = exp_arg_typ
-			got_arg_typ := c.check_expr_opt_call(arg.expr, c.expr(arg.expr))
+			mut got_arg_typ := c.check_expr_opt_call(arg.expr, c.expr(arg.expr))
 			node.args[i].typ = got_arg_typ
 			if no_type_promotion {
 				if got_arg_typ != exp_arg_typ {
@@ -1243,14 +1243,22 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				continue
 			}
 			if exp_arg_typ.has_flag(.generic) {
-				if concrete_types.len > 0 {
-					if utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
-						concrete_types)
-					{
-						exp_arg_typ = utyp
-					} else {
-						continue
-					}
+				if concrete_types.len == 0 {
+					continue
+				}
+
+				if exp_utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
+					concrete_types)
+				{
+					exp_arg_typ = exp_utyp
+				} else {
+					continue
+				}
+
+				if got_utyp := c.table.resolve_generic_to_concrete(got_arg_typ, method.generic_names,
+					concrete_types)
+				{
+					got_arg_typ = got_utyp
 				} else {
 					continue
 				}

--- a/vlib/v/checker/tests/generics_method_called_arg_mismatch.out
+++ b/vlib/v/checker/tests/generics_method_called_arg_mismatch.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_method_called_arg_mismatch.vv:3:15: error: cannot use `Foo` as `Bar` in argument 1 to `Obj.set`
+    1 | fn main() {
+    2 |     mut obj := Obj{}
+    3 |     obj.set<Bar>(Foo{})
+      |                  ~~~~~
+    4 | }
+    5 |

--- a/vlib/v/checker/tests/generics_method_called_arg_mismatch.vv
+++ b/vlib/v/checker/tests/generics_method_called_arg_mismatch.vv
@@ -1,0 +1,14 @@
+fn main() {
+	mut obj := Obj{}
+	obj.set<Bar>(Foo{})
+}
+
+struct Foo {}
+
+struct Bar {
+	Foo
+}
+
+struct Obj {}
+
+fn (mut o Obj) set<T>(val T) {}

--- a/vlib/v/tests/generics_with_nested_generics_fn_test.v
+++ b/vlib/v/tests/generics_with_nested_generics_fn_test.v
@@ -15,7 +15,7 @@ fn (ng NestedGeneric) nested_test<T>(mut app T) {
 
 fn method_test<T>(mut app T) int {
 	ng := NestedGeneric{}
-	ng.nested_test<T>(app)
+	ng.nested_test<T>(mut app)
 	return 22
 }
 

